hogben/utils.py

```diff
@@ -163,77 +157,182 @@ def fisher(qs: list[np.ndarray],
-                        for q, model in list(zip(qs, models))])
-
-    # Calculate the Fisher information matrix using equations from the paper.
-    M = np.diag(np.concatenate(counts) / r, k=0)
-    g = np.dot(np.dot(J.T, M), J)
-
-    # If there are multiple parameters,
-    # scale each parameter's information by its "importance".
-    if len(xi) > 1:
-        if isinstance(xi[0], refnx.analysis.Parameter):
-            lb = np.array([param.bounds.lb for param in xi])
-            ub = np.array([param.bounds.ub for param in xi])
-
-        elif isinstance(xi[0], bumps.parameter.Parameter):
-            lb = np.array([param.bounds.limits[0] for param in xi])
-            ub = np.array([param.bounds.limits[1] for param in xi])
-
-        # Scale each parameter with their specified importance,
-        # scale with one if no importance was specified.
-        importance_array = []
-        for param in xi:
-            if hasattr(param, "importance"):
-                importance_array.append(param.importance)
-            else:
-                importance_array.append(1)
-        importance = np.diag(importance_array)
+
+    def __init__(self,
+                 qs: list[np.ndarray],
+                 xi: list[Union['refnx.analysis.Parameter',
+                                'bumps.parameter.Parameter']],
+                 counts: list[int],
+                 models: list[Union['refnx.reflect.ReflectModel',
+                                    'refl1d.experiment.Experiment']],
+                 step: float = 0.005):
+        """Constructor for the Fisher matrix class.
+
+        Args:
+            qs: The Q points for each model.
+            xi: The varying model parameters.
+            counts: incident neutron counts corresponding to each Q value.
+            models: models to calculate gradients with.
+            step: step size to take when calculating gradient.
+            g: The Fisher information matrix
+            min_eigenval: The minimum eigenvalue associated with the Fisher
+             information matrix
+        """
+        self.qs = qs
+        self.xi = xi
+        self.counts = counts
+        self.models = models
+        self.step = step
+
+    @property
+    def n(self) -> int:
+        """The total amount of datapoints.
+
+        Returns:
+            int: total amount of datapoints.
+        """
+        return sum(len(q) for q in self.qs)
+
+    @property
+    def m(self) -> int:
+        """The total amount of parameters.
+
+        Returns:
+            int: total amount of parameters.
+        """
+        return len(self.xi)
+
+    @property
+    def g(self) -> np.ndarray:
+        """The Fisher information matrix.
+
+        Returns:
+            numpy.ndarray: The Fisher information matrix.
+        """
+        return self._calculate_g()
+
+    @g.setter
+    def g(self, value: np.ndarray):
+        """Sets the fisher information matrix."""
+        self.g = value
```

Comment:
I'm not fully up to speed with getters/setters, but should this not be `self._g = value`?

Reply:
Generally, the setter gets called whenever g is set. So `Fisher.g = value` calls the setter method and executes the code. As the getter method makes the attribute static, this is needed if you want to set the attribute directly. 

There's multiple things wrong with this method here:

1 - The value for g is never set directly in the code. This is probably not something we want to do anyway. From a user-facing perspective, this probably makes very little sense. And in the code itself, it's probably better to work with the hidden variable _g.
2 - Setting the property directly leads to an infinite loop. The setter is called whenever the value for `self.g` is set, since the setter directly sets `self.g` here, it calls itself ad infinitum. Hence the need to use `self._g` as you point out.
3 - Even if we use `self._g` here, as soon as we actually access the Fisher matrix, the getter method is called again, where self._g is redefined using the `_calculate_g` method. So whatever is set in the setter, gets overwritten as soon as g is actually accessed. 

```
self.g = [0,5,6] # Calls setter
b = self.g * 2 # Calls getter, completely undoing what is done in the line above
```

Either way, since we probably don't really want to change this property directly, but rather want it to be dependent on the model parameters etc... it probably doesn't make a lot of sense to use a setter in the first place. I was toying around with it in an earlier implementation, and in the end I didn't really need it. Am removing the setter from the code.